### ISSUE TITLE
@Immutable should be @Inherited and checked in sub classes.

### ISSUE THREAD

**Issue by Brixomatic** [spotbugs/spotbugs#1120]
Rationale:
```java 
javax.annotation.concurrent.Immutable
```

as it comes from:
```xml
<dependency>
  <groupId>com.google.code.findbugs</groupId>
  <artifactId>jsr305</artifactId>
  <version>3.0.2</version>
</dependency>
```
is not inherited.

It makes sense though, that you can annotate an abstract or non-final type as ```@Immutable``` to document developer intent and to have sub types checked for immutability by spotbugs or similar utilities. 
Depending on the circumstances, it might be beneficial to the design to have a class that is not "final", i.e. not immutable by design, but but immutable by contract. Alas Spotbugs cannot be used to warn from breaking that contract.

In this scenario:
```java
@Immutable
class X {
 final int x;
 public X(int x){this.x=x;}
 public String toString(){
  // reflectively create string from all fields (also if subtyped, bad design, but possible)
  return aString;
 }
}
class XY extends X{
 int y; // should be final to retain immutability
 public YX(int x, int y){
  super(x); 
  this.y=y;
 }
}
```
Spotbugs will not complain, even though the field y is not immutable.  
Still I could create a mutable Object "X" using a factory:
```java
public static X create(){ return new XY(1,2) };
```
Where the mutability would be visible in the toString() output.

**Comment by rovarga** [comment]
Well, neither of those annotations are `@java.lang.annotation.Inherited`, unlike `com.google.errorprone.annotations.Immutable`, hence I do not believe they are meant to extend the contract to subclasses. As per [JCIP docs](https://jcip.net/annotations/doc/net/jcip/annotations/Immutable.html) that version of `@Immutable` explicitly allows internal mutability, hence I do not believe SpotBugs should be reporting a warning in the case you highlighted -- it would lead to a ton of false positives.

**Comment by Brixomatic** [comment]
Sorry, I should have been more clear on this.

In order for an immutable type to be safely trusted as being immutable, any type that extends the immutable type must be immutable itself (Liskovs Substitution Principle is debatable, but in this case it does hint in the same direction), else it would break the @Immutable contract of the parent type. 

So there should be either a warning that class XY is not marked @Immutable, or it must be checked like any other immutable class, hence @Immutable should be inherited.

In this case, the field "y" is non-final _and_ non - private, regardless of any lazy init, its mutability cannot be not internal only. The field should either be final or private. Yes, this still leaves non - private accessors open and any secure check for immutability is crazy complex, but being final or private is a minimum requirement for being immutable.

**Comment by Brixomatic** [comment]
No more investigation on this?

**Comment by rovarga** [comment]
> No more investigation on this?

I agree with your assessment. As far as I understand (in-brain cache, obviously not reliable), though, SB treats all .Immutable contracts the same -- and it should not.

Addressing this issue may involve quite a bit of surgery on SB internals -- for example to differentiate the different levels of immutability and the corresponding rules as to what is okay and what's not.

Note, that with advent of [nest-mates](https://openjdk.java.net/jeps/181) even 'private' visibility is not enough to guarantee access. That brings us into heuristics-vs-trade-offs territory -- and I my views on that do not align with those of SB maintainers, so I prefer not to go there :)
